### ISSUE TITLE
Intermittent `WriteMessagesFailed` errors during periods of high activity

### ISSUE THREAD

**Issue by lucavice** [akkadotnet/Akka.Persistence.Linq2Db#430]
**Version Information**
Version of Akka.NET? 1.5.25
Which Akka.NET Modules? Akka.NET, Akka.Persistence, Akka.Persistence.Sql

**Describe the bug**
We have now been running [Akka.Persistence.Sql](https://github.com/akkadotnet/Akka.Persistence.Sql) on our prod environment for a few weeks. Once a week, we run an import process that writes events fairly quickly in the journal.  Things seem to work ok, database telemetry shows max usage peaking at around 15%, but once every few thousands writes, we suddenly get an exception from the actor message `Akka.Persistence.WriteMessagesFailed`:

```
Message.Cause.Message: Sequence contains no elements
Message.Cause.Source: System.Linq
Message.Cause.StackTrace:
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at Akka.Persistence.Sql.Journal.SqlWriteJournal.WriteMessagesAsync(IEnumerable`1 messages)
   at Akka.Util.Internal.AtomicState.CallThrough[T,TState](TState state, Func`2 task)
   at Akka.Util.Internal.AtomicState.CallThrough[T,TState](TState state, Func`2 task)
   at Akka.Persistence.Journal.AsyncWriteJournal.ExecuteBatch(WriteMessages message, Int32 atomicWriteCount, IActorRef self, Int64 resequencerCounter)
```

This seems to be quite a low-level error, so that's why I am reporting it here.
This error also causes all subsequent event replays to fail for 2 minutes with:

```
Akka.Pattern.OpenCircuitException: Circuit Breaker is open; calls are failing fast
   at Akka.Persistence.Journal.AsyncWriteJournal.<>c__DisplayClass18_0.<<HandleReplayMessages>g__ExecuteHighestSequenceNr|0>d.MoveNext()
```

The OpenCircuitException is particularly problematic, as it effectively stops most activity in the system for 2 minutes.

Unfortunately, I am not sure how to replicate this. It happened 3 times this morning over the course of a few hours, as we were importing about 20K orders in our system.

The exact same process did not raise these exceptions while we were running `Akka.Persistence.SqlServer` and Akka.NET v1.4.

I'm wondering if this is an actual bug, or if there are any settings that we can play with to prevent it, make it less likely, or any sort of advice that we can follow.

We have not enbled yet the WriterUuid Anti-Corruption Layer Feature, nor the tag table.

Thank you so much!

EDIT: Once, I also got the following exception:
```
System.TimeoutException: The operation has timed out.
   at Akka.Util.Internal.AtomicState.CallThrough[T,TState](TState state, Func`2 task)
   at Akka.Util.Internal.AtomicState.CallThrough[T,TState](TState state, Func`2 task)
   at Akka.Persistence.Journal.AsyncWriteJournal.<>c__DisplayClass18_0.<<HandleReplayMessages>g__ExecuteHighestSequenceNr|0>d.MoveNext()
```
It did not cause OpenCircuitException to trigger, and I'm not sure if it's a separate bug or what, but I'm adding it here in case it provides additional hints and context to the original issue.

**Comment by to11mtm** [comment]
Off the top of my head, this could happen if all writes submitted to the
journal were 'rejected' and we sent an empty set down the stream...

This may be a regression, maybe not (thought we had a guard for this...)

Will try to peek today if I have time

**Comment by to11mtm** [comment]
Okay, so these are the 'possibilities' that come to mind:

Here, https://github.com/akkadotnet/Akka.Persistence.Sql/blob/644bdebdb9d486d8b092da2eb3822f6faf1205f4/src/Akka.Persistence.Sql/Journal/SqlWriteJournal.cs#L210-L229 the method that your callstack is showing

One could see a throw on `.Head()` here, but that would require us to somehow get an empty IEnumerable (which I believe is against contract.) I'd be surprised if this actually happened *normally*, but this could totally happen if you are partying on `SqlWriteJournal` in a weird way, i.e. manually instantiating it and just calling `WriteMessagesAsync` (Doesn't look like it here but...) There's a few things it does under the assumption it's working under a normal Persistent actor lifecycle for writes/etc.

The other possibility, is that somewhere in DAO `AsyncWriteMessages` https://github.com/akkadotnet/Akka.Persistence.Sql/blob/644bdebdb9d486d8b092da2eb3822f6faf1205f4/src/Akka.Persistence.Sql/Journal/Dao/BaseByteArrayJournalDao.cs#L94C1-L109C10 we have some edge case that wasn't forseen on the write queue. The 'edge case' here would be the one I mentioned before:

> Off the top of my head, this could happen if all writes submitted to the
journal were 'rejected' and we sent an empty set down the stream...

But I know we have guard code in the write pipeline for a lot of that, so I'd still be surprised TBH.

-----

May have other thoughts, will keep pondering in meantime...

**Comment by lucavice** [comment]
Thank you so much @to11mtm for looking into this.

> I'd be surprised if this actually happened normally, but this could totally happen if you are partying on SqlWriteJournal in a weird way, i.e. manually instantiating it and just calling WriteMessagesAsync

Just to clarify, the only way I interact with the SqlWriteJournal is through standard `Persist` calls in standard Persitent Actors, no manual instantiation or other shenenigans. 

Looking more deeply into our logs I found that the problem is generated in a specific condition: whenever the importer processes an order with a particularly high number of order items. This causes the system to start up a lot of persistent actors in a very short time frame. Then, once a single `WriteMessagesFailed` shows up, there are many other that follows (I counted 18 in one intance) across many different persistent actors. This is very good news because it means that I will probably be able to replicate this reliably on our staging environment and, if I'm lucky, even on local development environment. In that case, I'll be able to peak into what's happening and confirm one of your scenarios.

This also reminded me of a very similar issue I had about a couple of years ago that I also opened an issue for: https://github.com/akkadotnet/akka.net/issues/6355. I never followed up much on the original issue because it seemed to disappear once we moved to the SQL batching journal (and unfortunately, I didn't have much time to dig deeper). If I remember correctly, even that issue was raised in a very similar circumstance. I actually have a strong suspicion that this issue is a reincarnation of that, since we haven't changed the relevant code since then (a potential misuse of `DeferAsync` inside `AroundReceive`). Even in that issue the call stack pointed to something within `WriteMessagesAsync(IEnumerable`1 messages)` gone wrong, so there seems to be a lot of similarities.

I even wrote a repro to the original issue, so I should even be able to swap the write plugin to `Akka.Persistence.Sql` and see if it comes up.

So @to11mtm, I think for now I've got many new elements to investigate, I will report here any new findings.

**Comment by to11mtm** [comment]
> This causes the system to start up a lot of persistent actors in a very short time frame

That can have it's own challenges, although in my observation it is -usually- on the number of readers, since we have (what should normally be) a fixed number of writers at (by default) a fairly low number.

With `Akka.Persistence.Sql` I would expect excessive readers case to error out in either as an -explicit- failure on WriteMessages (you'd see the Connection timeout on the writes vs what you are seeing) or... noting the edit I now see, a timeout error on reads of sequences or messages... 

*But*, if connection this was expressing on the write side, I'd expect to see timeout errors on the write side as well, unless there's something I don't know about (yet).

After that disclaimer, crazy question that may help your investigation, are you able to try adjusting the `Max Pool Size` on the ADO connection? Default is 100, try increasing it to 200 and see if that makes things better?

**Comment by lucavice** [comment]
Hey @to11mtm, it turned out that it was related to https://github.com/akkadotnet/akka.net/issues/6355 and caused by an illegal use of `DeferAsync` inside a `AroundReceive`. I only had to tweak the original bug repro code with Akka 1.5 and Akka.Persitence.Sql and I got the same exception. If you are interested and want to run it locally, the full repro is in a small self-contained console app is here: https://github.com/lucavice/AkkaNetBugRepro/tree/akka-1.5-with-persistence-sql

At the time @Aaronontheweb already mentioned that using a `DeferAsync` inside a `AroundReceive` was not supported and an improper use of it. Our use case was implementing a category of actors that offer an automatic confirmed delivery message back to the sender after the events are persisted (for the legacy at-least-once-delivery actor). Anyway, I still have no idea why it fails the way it does and how it's related to the DeferAsync. If I comment out that line, the issue disappears.
It does seem that it's breaking the contract and sending an emtpy IEnumerable. The stack trace outputted by the repro console app clearly indicates a failure at the `Head` method that you mentioned:

```
WriteMessagesFailed received: System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at LanguageExt.List.head[T](IEnumerable`1 list)
   at ListExtensions.Head[T](IEnumerable`1 list)
   at Akka.Persistence.Sql.Journal.SqlWriteJournal.WriteMessagesAsync(IEnumerable`1 messages)
   at Akka.Persistence.Journal.AsyncWriteJournal.<>c.<ExecuteBatch>b__21_0(ValueTuple`2 state)
   at Akka.Util.Internal.AtomicState.CallThrough[T,TState](TState state, Func`2 task)
```

The solution for me is to refactor our code to remove that problematic DeferAsync, but I'm still kind of interested to know what's happening under the hood and why it fails the way it does.

Anyway, thank you again for all the help, really appreciated.
